File: LeetCode.com/103-BinaryTreeZigzagLevelOrderTraversal.cs
//103. Binary Tree Zigzag Level Order Traversal

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    //IList<IList<int>> oList = new List<List<int>>();

    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
     
        List<IList<int>> res = new List<IList<int>>();
        
        if(root == null)
            return res;
        
        Queue<TreeNode> queue = new Queue<TreeNode>();
        queue.Enqueue(root);
        
        int rowNum = 0;
        while(queue.Count > 0) {
            List<int> row = new List<int>();
            int count = queue.Count;
            
            for(int i = 0; i < count; i++)
            {
                TreeNode node = queue.Dequeue();
                row.Add(node.val);
                
                if(node.left != null)
                    queue.Enqueue(node.left);
                if(node.right != null)
                    queue.Enqueue(node.right);
            }
            
            if(rowNum % 2 == 1)
                row.Reverse();
                
            res.Add(row);
            
            rowNum++;
        }
            
        return res;
    }
    
    
    
}

File: LeetCode Java/14 Patterns/BFS/103-BinaryTreeZigZagLevelOrderTraversal.java
/*
 * 103. Binary Tree Zigzag Level Order Traversal
 * Given the root of a binary tree, return the zigzag level order traversal of
 * its nodes' values. (i.e., from left to right, then right to left for the next
 * level and alternate between).
 * Example 1:
 * Input: root = [3,9,20,null,null,15,7]
 * Output: [[3],[20,9],[15,7]]
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {

        Queue<TreeNode> q = new LinkedList<>();
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (root == null)
            return res;
        q.add(root);

        boolean revFlag = false;

        while (!q.isEmpty()) {
            var size = q.size();
            List<Integer> level = new ArrayList<>();

            for (int i = 0; i < size; i++) {
                var curr = q.remove();
                level.add(curr.val);
                if (curr.left != null) { q.add(curr.left); }
                if (curr.right != null) { q.add(curr.right); }
            }
            if (revFlag) {
                Collections.reverse(level);
                revFlag = false;
            } else {
                revFlag = true;
            }
            res.add(level);
        }
        return res;
    }
}